batcheval: refactor split trigger stats computation

tbg opened the pull request:
Pulling out the stats-related computations into a helper taylored
exactly to what happens during a split provides a convenient venue for
lengthy rumination on the finer points of said stats computations, while
decluttering the split trigger's main method.

This was motivated by #37583, which needs to adjust some of this code.
Now that it's (hopefully) more accessible, it should be more straight-
forward to figure out what exactly to do.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38084)
<!-- Reviewable:end -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38084)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @giorgosp, @nvanbenschoten, and @tbg)

___
*[pkg/storage/batcheval/cmd_end_transaction.go, line 860 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzVrCOAEzR5Mdx6qXR:-LgzVrCOAEzR5Mdx6qXS:b-eb2yu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/cmd_end_transaction.go#L860)):*
> ```Go
> }
> 
> func splitTriggerHelper(
> ```

It's not clear to me what value this helper provides.
___
*[pkg/storage/batcheval/split_stats_helper.go, line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzWXwZBI__H2dX5Or3:-LgzWXwZBI__H2dX5Or4:b-cbb86u) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/split_stats_helper.go#L44)):*
> ```Go
> //   DeltaBatch contained in contributions attributable to the keyspace on the
> //   left).
> // - CombinedError: the difference between (AbsPreSplitBoth+DeltaBatch) and the
> ```

Consider renaming this to `CombinedErrorDelta` so it's clear that this isn't an `error`.
___
*[pkg/storage/batcheval/split_stats_helper.go, line 101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzYUS48HiQIgkAio1g:-LgzYUS51dM3SyW153-w:boi0meu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/split_stats_helper.go#L101)):*
> ```Go
> // recompute AbsPostSplitRight from scratch. (As fallout, we can in principle
> // compute CombinedError, but we don't care).
> type SplitStatsHelper struct {
> ```

Do all of these types and functions need to be exported?
___
*[pkg/storage/batcheval/split_stats_helper.go, line 110 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzXxxK5S10KOz2svt9:-LgzXxxK5S10KOz2svtA:b-b54j1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/split_stats_helper.go#L110)):*
> ```Go
> type SplitStatsHelperInput struct {
> 	AbsPreSplitBothEstimated enginepb.MVCCStats
> 	DeltaBatchEstimated      enginepb.MVCCStats
> ```

If we had a `DeltaBatchRHSEstimated` could we avoid the need for a mutable `AbsPostSplitRight()`? Is there a benefit of this approach?
___
*[pkg/storage/batcheval/split_stats_helper.go, line 112 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzWzZA2Rp-iRHDhcOh:-LgzWzZA2Rp-iRHDhcOi:b-o23um0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/split_stats_helper.go#L112)):*
> ```Go
> 	DeltaBatchEstimated      enginepb.MVCCStats
> 	AbsPostSplitLeft         enginepb.MVCCStats
> 	AbsPostSplitRightFn      func() (enginepb.MVCCStats, error)
> ```

You say it in the comment above, but I'd note again here that we try not to need to call this.
___
*[pkg/storage/batcheval/split_stats_helper.go, line 154 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzXJBu7dwQwFZr87m2:-LgzXJBu7dwQwFZr87m3:b67e6p1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/split_stats_helper.go#L154)):*
> ```Go
> func (h SplitStatsHelper) AbsPostSplitRight() *enginepb.MVCCStats {
> 	return h.absPostSplitRight
> ```

Extra line.


<!-- Sent from Reviewable.io -->

giorgosp reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-:-Lh4Xv6N78lyfdFYMl1c:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @giorgosp and @tbg)

___
*[pkg/storage/batcheval/split_stats_helper.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-Lh4X8Zb8fm08uw2ItKR:-Lh4X8Zb8fm08uw2ItKS:b-hr8csr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/split_stats_helper.go#L24)):*
> ```Go
> //   i.e. without accounting for any writes in the batch. This can have
> //   ContainsEstimates set.
> // - DeltaBatchEstimated: the writes in the batch. Since we have no reason to
> ```

I have a question trying to understand it more deeply: What does `DeltaBatchEstimated` actually contain? The aggregated diff between new stats (from batch writes) and current stats? Also, "batch writes" contain the stats just for the range that we are going to split?


<!-- Sent from Reviewable.io -->

giorgosp reviewed (approved):
> Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38084)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38084)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @giorgosp and @nvanbenschoten)

___
*[pkg/storage/batcheval/cmd_end_transaction.go, line 860 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzVrCOAEzR5Mdx6qXR:-LhHF2xhA1FQFNQ9DBll:b-skumj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/cmd_end_transaction.go#L860)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's not clear to me what value this helper provides.
</blockquote></details>

I want to make sure there aren't any other stats-y objects in scope. The helper sees only the stats helper and not, for example, `bothDeltaMS` or `leftMS`.
___
*[pkg/storage/batcheval/split_stats_helper.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-Lh4X8Zb8fm08uw2ItKR:-LhHIHugES4ll5oK399x:b8073q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/split_stats_helper.go#L24)):*
<details><summary><i>Previously, giorgosp (George Papadrosou) wrote…</i></summary><blockquote>

I have a question trying to understand it more deeply: What does `DeltaBatchEstimated` actually contain? The aggregated diff between new stats (from batch writes) and current stats? Also, "batch writes" contain the stats just for the range that we are going to split?
</blockquote></details>

It's the effects of evaluating the EndTransaction (without accounting for the split). So it'll have a write to the txn record, and resolving the intent on the range descriptor (which is local to this range, so it's resolved synchronously with the commit - this is actually important for various reasons). I tried to clarify the comment.
___
*[pkg/storage/batcheval/split_stats_helper.go, line 101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzYUS48HiQIgkAio1g:-LhHIv6TD0VQhG1pxUjR:bknfm1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/split_stats_helper.go#L101)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do all of these types and functions need to be exported?
</blockquote></details>

There are two competing philosophies here:

1. don't export anything that doesn't need to be exported (and nothing here needs to be exported) and
2. export the "public API" to distinguish it from the internal one

I opted for 2) to make it clear what the intended usage is, as a way to compensate for the lack of an equivalent of private member variables. For example, you're not supposed to alter `SplitStatsHelper.in` after you've created the helper. Also exporting the API forces it to be linted and makes it show up in godoc (not that anyone uses this). And it'll be clear how to pull this into its own package (perhaps with surrounding splits-related code) if we want to do that in the future.

WDYT?
___
*[pkg/storage/batcheval/split_stats_helper.go, line 110 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzXxxK5S10KOz2svt9:-LhHKIEM378trZYE-9nL:b-d5ipn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/split_stats_helper.go#L110)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we had a `DeltaBatchRHSEstimated` could we avoid the need for a mutable `AbsPostSplitRight()`? Is there a benefit of this approach?
</blockquote></details>

Maybe I don't understand the suggestion. To get `DeltaBatchRHSEstimated` we'd have to bin all writes into left/right all throughout the `EndTransaction` evaluation, which can be done but seems worse. Let's assume we did that, then we'd still have to move any writes to the RHS to before the helper is created. But we have this code here which needs to know the current stats and returns a mutated one:

https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/cmd_end_transaction.go#L1014-L1019

I also don't have a problem with writing into `*AbsPostSplitRight()`. In the method where this is done this is the only writable stats field, and we've decreed that only writes to the RHS can occur, so this is very similar to having a `*MVCCStats` in scope. We could also make one and pass it into the helper, but then that looks even more magical because it's not obvious that the two are connected except on the line that creates the helper.

What we could do of course is to have


<!-- Sent from Reviewable.io -->

tbg committed:
batcheval: refactor split trigger stats computation

Pulling out the stats-related computations into a helper taylored
exactly to what happens during a split provides a convenient venue for
lengthy rumination on the finer points of said stats computations, while
decluttering the split trigger's main method.

This was motivated by #37583, which needs to adjust some of this code.
Now that it's (hopefully) more accessible, it should be more straight-
forward to figure out what exactly to do.

Release note: None

tbg commented:
Oops, never finished a comment there:

> What we could do of course is to have

...

a local MVCCStats that you need to pass into both of the methods that give you a computed MVCCStats back.

Anyway, I'll leave this as is for now. I'm going to go ahead and merge this to unblock @giorgosp, but happy to continue the discussion and to send a follow-up PR.

bors r=giorgiosp

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38084)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @giorgosp, @nvanbenschoten, and @tbg)

___
*[pkg/storage/batcheval/split_stats_helper.go, line 101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzYUS48HiQIgkAio1g:-LhMFfkGDWa8FqFZYJVe:bjcj7ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/split_stats_helper.go#L101)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

There are two competing philosophies here:

1. don't export anything that doesn't need to be exported (and nothing here needs to be exported) and
2. export the "public API" to distinguish it from the internal one

I opted for 2) to make it clear what the intended usage is, as a way to compensate for the lack of an equivalent of private member variables. For example, you're not supposed to alter `SplitStatsHelper.in` after you've created the helper. Also exporting the API forces it to be linted and makes it show up in godoc (not that anyone uses this). And it'll be clear how to pull this into its own package (perhaps with surrounding splits-related code) if we want to do that in the future.

WDYT?
</blockquote></details>

I agree about exporting the public API of the type, but the type itself (and its constructor) doesn't need to be exported. There are no users or it outside of the `batcheval` package and it certainly isn't part of the `batcheval` package's API.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38084)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @giorgosp, @nvanbenschoten, and @tbg)

___
*[pkg/storage/batcheval/cmd_end_transaction.go, line 860 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38084#-LgzVrCOAEzR5Mdx6qXR:-LhMGUrx-3n8G_JsajHo:b-r7qe9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b0b7fdf76e16b3681dc63136d2df875bd086c81/pkg/storage/batcheval/cmd_end_transaction.go#L860)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I want to make sure there aren't any other stats-y objects in scope. The helper sees only the stats helper and not, for example, `bothDeltaMS` or `leftMS`.
</blockquote></details>

Ok, then do you mind adding a comment saying as much? 


<!-- Sent from Reviewable.io -->
